docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+**Problem**
+
+We've got at least 3 cases where kube needs to know about and interact with functionality provided by the underlying infrastructure. The 3 I can think of from the top of my head are:
+
+1. minion discovery
+2. storage (be it a GCE PD, NFS, Ceph, or something else [durable data is a little special])
+3. externally reachable functionality (external IPs, external load balancers, ssl termination points, etc)
+
+We don't seem to have a common pattern how kube and the cloud provider/infrastructure should be working together to expose functionality. I'd like to start a discussion of how I see at least a start of how external functionality and kube can get along.
+
+**The three ways kube and infrastructure interact**
```

Comment:
I think of Kubernetes as a layered system:

At the bottom, we need an infrastructure orchestration layer, which can provision compute, storage, and network resources on demand, in environments that support that, such as public and private clouds.  That should be the cloudprovider API, #2770.

Above that, we need a (hopefully minimal) infrastructure configuration management layer, such as for configuring OS images. Currently we use Salt for this, but I'd like this to be pluggable (Salt, Fleet, Ansible, Puppet, Chef, Bosh, ...), or subsumed by Kubernetes. At least all application configuration (Kubelet, Docker, ...) should be self-hosted. For networking, we've talked about using Flannel or Weave.

The thing we think of as Kubernetes is an application-centric scheduling, resource management, and orchestration system. As much as possible, it should provision the underlying resources on demand. For instance, if someone requests pods with large resource requirements or a large number of pods, Kubernetes should be able to request resources that would satisfy those requests from the underlying infrastructure layer. This also applies to custom configuration of existing resources, such as reconfiguring nodes to be security hardened or reserving them for sole tenancy on demand. I call this "pull".

Kubernetes should also support external "push" of resources into the system, by an administrator or arms-length automatic provisioning system, such as a VM auto-scaler. These resources need to be modeled so that they can be allocated by Kubernetes. The only case of this today is minions/nodes, but we've also discussed doing this for IP addresses and storage volumes.

Supporting attachment to external resources is also reasonable. 

I agree we need all 3 approaches.

Not as relevant to this discussion is layers above or adjacent to Kubernetes, such as Openshift and other PaaSes, but we should document our desired approach, which is composition of complementary, a la carte functionality.